### PR TITLE
feat: allow for default values in contract exported functions

### LINKED ISSUE #34: exported functions can not have default parameter values
this error is raised when calling the `customize` function below because this is the **entry function to the contract**

```bash
Scheduling a call: dev-1586182981781.customize()
Error:  TypedError: Transaction FCpyq6UJS73rzEHS9jURr4dcF97SqFLmSdZy1YktS5zR failed. 
        String encoding is bad UTF-16 sequence.
```

from @willemneal 
> _The exported functions cannot have default values. You could just make a default function that calls the customize function.  If you add `//@nearfile out` at the top the file, when compiling you can see the code output that is about to compiled. Every exported function get’s a helper `__wrapper` function that deserializes the inputs and then calls the function. As of now this wrapper function requires that you have all arguments in the JSON Object._

below is the function that throws the error

```ts
/**
 * This function supports the customization of this ERC-20 token before initialization
 * It may or may not be called.  If not called, the contract uses sensible defaults
 * If called, it can only be called once (!) and prevents repeat calls
 *
 * NOTE: this function uses storage keys with _underscore prefix since these are guaranteed not
 * to conflict with accounts on the NEAR platform. see https://nomicon.io/DataStructures/Account.html#examples
 *
 * THIS IS NOT part of the ERC-20 spec
 *
 * @param name of the token
 * @param symbol for the token
 * @param decimals places used when rendering the token
 * @param supply of the tokens in total at launch
 */
export function customize(
  name: string = "Solidus Wonder Token",        // awesome name for a token
  symbol: string = "SWT",                       // pronounced "sweet", rhymes with "treat"
  decimals: u8 = 2,                             // number of decimal places to assume for rendering,
  supply: u128 = u128.from(100_000_000),        // <raised pinky> one meeeeellion coins ... divisible in 100ths,
  exchangeRate: u8 = 100                        // of these ERC-20 tokens per NEAR token
  ) : void {
  DEBUG ? logging.log("[call] customize('" + name + "', '" + symbol + "', " + decimals.toString() + ", " + supply.toString() + ", " + exchangeRate.toString() + ")") : false;
  const owner = assertTrueOwner()
  // only set values that are provided, otherwise ignore
  storage.setString("_bank", owner);
  storage.setString("_name", name);
  storage.setString("_symbol", symbol);
  storage.set<u8>("_decimals", decimals);
  storage.set<u128>("_totalSupply", supply);
  storage.set<u8>("_exchangeRate", exchangeRate);
}
```

### PR DESCRIPTION
**Pull request by willemneal**
~Still need to add check to ensure that no non-initialized parameter is after the first default parameter~

On second thought this shouldn't matter since the arguments are coming from JSON and any missing without an initializer will fail.

fixes #34

### CODE REVIEW

File: assembly/__tests__/runtime/main.ts [hunk 1]
```diff
@@ -557,3 +557,10 @@ export function setValue(value: string): string {
 export function getValue(): string | null {
   return storage.get<string>("name");
 }
+
+export function getDefaultValue(y: string = "hello world", x: i32 = 42): i32 {
+  assert(x == 42, "x is not default value of 42");
+  assert(y == "hello world", "y is not default value of hello world");
```
Comment (janedegtiareva) [inline assembly/__tests__/runtime/main.ts:563]: Hello world is not default value of y?

File: assembly/__tests__/runtime/main.ts [hunk 2]
```diff
@@ -557,3 +557,10 @@ export function setValue(value: string): string {
...
+  logging.log(y);
```
Comment (janedegtiareva) [inline assembly/__tests__/runtime/main.ts:564]: nit: this is redundant with the assert

File: bindgen/src/JSONBuilder.ts [hunk 3]
```diff
@@ -148,7 +148,16 @@ export class JSONBindingsBuilder extends BaseVisitor {
...
     if (params.length > 0) {
       this.sb[this.sb.length - 1] += params
-        .map((param) => createDecodeStatement(param))
+        .map((param) => {
+          let name = toString(param.name);
+          let type = toString(param.type);
+          let res = `obj.has('${name}') ? 
+             ${createDecodeStatement(param)} : 
+             assertNonNull<${type}>('${name}', <${type}>${
+            param.initializer ? toString(param.initializer) : "null"
```
Comment (janedegtiareva) [inline bindgen/src/JSONBuilder.ts:157]: looks like there are lint issues. is the linter checking this file?

File: assembly/__tests__/runtime/main.ts (hunk 1 above)
Comment (willemneal) [reply assembly/__tests__/runtime/main.ts:563]: It is.  This is a test that when compiled and the default values still work when the function is called in the nearcore test.

File: assembly/__tests__/runtime/main.ts (hunk 2 above)
Comment (willemneal) [reply assembly/__tests__/runtime/main.ts:564]: Lol true, but it was just to see that when the test runs it prints `hello world`.  I can remove it though.

File: bindgen/src/JSONBuilder.ts (hunk 3 above)
Comment (willemneal) [reply bindgen/src/JSONBuilder.ts:157]: Yeah, but I'm not sure if it checks template strings.  I did this for readability in the generated source code.

File: assembly/__tests__/runtime/main.ts (hunk 2 above)
Comment (willemneal) [reply assembly/__tests__/runtime/main.ts]: Removed it ;-)

File: assembly/__tests__/runtime/main.ts [hunk 4]
```diff
@@ -557,3 +557,9 @@ export function setValue(value: string): string {
 export function getValue(): string | null {
   return storage.get<string>("name");
 }
+
+export function getDefaultValue(y: string = "hello world", x: i32 = 42): i32 {
+  assert(x == 42, "x is not default value of 42");
```
Comment (janedegtiareva) [inline assembly/__tests__/runtime/main.ts:562]: 42 is not the default value of x?

Comment (willemneal) [reply assembly/__tests__/runtime/main.ts:562]:
```suggestion
  assert(x == 42, "expected default of 42, got: " + x.toString());
```

**Pull request merged**